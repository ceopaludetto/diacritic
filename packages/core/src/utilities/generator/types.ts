import type { ResourceGraph } from "../resource";
import type { DiacriticGenerationOptions, Entry, Parser } from "../types";

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { EOL } from "node:os";
import { dirname } from "node:path";

import { dset } from "dset";

import { prefixes } from "../loader";

/// keep-sorted
type GenerateTypesOptions = {
	defaultLanguage: string;
	generation: DiacriticGenerationOptions;
	languages: string[];
	namespaces: string[];
	parser: Parser;
	resourceGraph: ResourceGraph;
};

function toUnion(arr: string[]) {
	return arr.map(item => `"${item}"`).join(" | ");
}

function functionFromEntry(entry: Entry) {
	return `(${entry.args.map(item => item.name + ": " + item.type)}) => string`;
}

export function generateTypes({
	defaultLanguage,
	generation,
	languages,
	namespaces,
	parser,
	resourceGraph,
}: GenerateTypesOptions) {
	const entries = resourceGraph.allEntriesForLanguage(defaultLanguage);
	const declarations: string[] = generation?.banner ?? [];

	declarations.push(
		`declare module "${prefixes[1]}registry" {`,
		`\texport const defaultLanguage: ${JSON.stringify(defaultLanguage)};`,
		`\texport const languages: (${toUnion(languages)})[];`,
		`\texport const namespaces: (${toUnion(namespaces)})[];`,
		"",
		`\texport type SupportedLanguages = typeof languages;`,
		`\texport type SupportedNamespaces = typeof namespaces;`,
		`}`,
		`declare module "${prefixes[2]}registry" {`,
		`\texport * from "${prefixes[1]}registry";`,
		`}`,
	);

	declarations.push(
		`declare module "${prefixes[1]}proxy" {`,
		`\ttype Proxy = {`,
	);

	for (const [namespace, files] of Object.entries(entries)) {
		const contents = files.map(file => readFileSync(file, "utf-8")).flatMap(parser.convertFile);
		const structure = {};

		for (const entry of contents) dset(structure, entry.path, functionFromEntry(entry));

		declarations.push(
			`\t\t${namespace}: ${JSON.stringify(structure).replace(/"/g, "")};`,
		);
	}

	declarations.push(
		`\t};`,
		`}`,
		`declare module "${prefixes[2]}proxy" {`,
		`\texport * from "${prefixes[1]}proxy";`,
		`}`,
	);

	const folder = dirname(generation.outFile);
	if (!existsSync(folder)) mkdirSync(folder, { recursive: true });

	writeFileSync(generation.outFile, declarations.join(EOL), "utf-8");
}
