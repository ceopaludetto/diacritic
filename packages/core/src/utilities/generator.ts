import type { ResourceGraph } from "./resource";
import type { diacriticGenerationOptions, Entry, Parser } from "./types";

import { readFileSync, writeFileSync } from "node:fs";
import { EOL } from "node:os";

import { prefixes } from "./loader";

export function createFunctionFromEntry(interpolation: Parser["interpolation"], entry: Entry) {
	const interpolators = interpolation.map(side => side.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"));
	const regex = new RegExp(`${interpolators[0]}\\s*(\\w+)\\s*\\|\\s*[\\w\\s]*\\s*${interpolators[1]}`, "g");

	return [
		`export function ${entry.name}(${entry.args.map(item => item.name + ": " + item.type)}) {`,
		`\treturn \`${entry.return.replace(regex, (_, content) => `\${${content}}`)}\`;`,
		`}`,
	].join(EOL);
}

export function createExportFromLanguageAndNamespace(language: string, namespace: string) {
	return `export * as ${namespace} from "${prefixes[0]}${language}/${namespace}";`;
}

export function createImportFromLanguageAndNamespace(language: string, namespace: string) {
	return `import * as ${language} from "${prefixes[0]}${language}/${namespace}";`;
}

export function createConditionalFunctionFromEntry(defaultLanguage: string, languages: string[], entry: Entry) {
	const cases = languages
		.flatMap((language, index) => {
			const fallback = language !== defaultLanguage
				? `?? ${defaultLanguage}?.${entry.name}?.(${entry.args.map(item => item.name).join(", ")})`
				: "";

			return [
				index === 0 ? `\t\tcase "${language}":` : `\t\tcase "${language}":`,
				`\t\t\treturn ${language}?.${entry.name}?.(${entry.args.map(item => item.name).join(", ")})${fallback}`,
			];
		})
		.join(EOL);

	return [
		`export function ${entry.name}(${entry.args.map(item => item.name + ": " + item.type)}) {`,
		`\tswitch (r.language()) {`,
		`${cases}`,
		`\t}`,
		`}`,
	].join(EOL);
}

export function createImportExport(imports: string[], exports: string[]) {
	return [...imports, "", ...exports].join(EOL);
}

export function createTypesFromEntry(entry: Entry) {
	return `export function ${entry.name}(${entry.args.map(item => item.name + ": " + item.type).join(", ")}): string;`;
}

export function createRegistry(defaultLanguage: string, languages: string[]) {
	return [
		`export const defaultLanguage: string = ${JSON.stringify(defaultLanguage)};`,
		`export const languages: string[] = ${JSON.stringify(languages)};`,
	].join(EOL);
}

/// keep-sorted
type GenerateTypesOptions = {
	defaultLanguage: string;
	generation: diacriticGenerationOptions;
	parser: Parser;
	resourceGraph: ResourceGraph;
};

export function generateTypes({ defaultLanguage, generation, parser, resourceGraph }: GenerateTypesOptions) {
	const entries = resourceGraph.allEntries();
	const declarations: string[] = generation?.banner ?? [];

	for (const [language, namespaces] of Object.entries(entries)) {
		for (const [namespace, files] of Object.entries(namespaces)) {
			const content = files.map(file => readFileSync(file, "utf-8"));

			const functions = content.flatMap(item => parser.convertFile(item)).map(item => "\t" + createTypesFromEntry(item));

			declarations.push(
				`declare module "${prefixes[1]}${language}/${namespace}" {`,
				`${functions.join(EOL)}`,
				`}`,
				`declare module "${prefixes[2]}${language}/${namespace}" {`,
				`\texport * from "${prefixes[1]}${language}/${namespace}";`,
				`}`,
			);

			if (language === defaultLanguage) {
				declarations.push(
					`declare module "${prefixes[1]}${namespace}" {`,
					`\texport * from "${prefixes[1]}${language}/${namespace}";`,
					`}`,
					`declare module "${prefixes[2]}${namespace}" {`,
					`\texport * from "${prefixes[1]}${namespace}";`,
					`}`,
				);
			}
		}

		const exports = Object.keys(namespaces).map(namespace => `\texport * as ${namespace} from "${prefixes[1]}${language}/${namespace}";`);
		declarations.push(
			`declare module "${prefixes[1]}${language}" {`,
			`${exports.join(EOL)}`,
			`}`,
			`declare module "${prefixes[2]}${language}" {`,
			`\texport * from "${prefixes[1]}${language}";`,
			`}`,
		);
	}

	writeFileSync(generation.outFile, declarations.join(EOL), "utf-8");
}
