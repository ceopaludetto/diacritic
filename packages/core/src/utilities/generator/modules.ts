import type { Entry, Parser } from "../types";

import { EOL } from "node:os";

import { prefixes } from "../loader";

export function createFunctionFromEntry(interpolation: Parser["interpolation"], entry: Entry) {
	const interpolators = interpolation.map(side => side.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"));
	const regex = new RegExp(`${interpolators[0]}\\s*(\\w+)\\s*\\|\\s*[\\w\\s]*\\s*${interpolators[1]}`, "g");

	return [
		`export function ${entry.name}(${entry.args.map(item => item.name + ": " + item.type)}) {`,
		`\treturn \`${entry.return.replace(regex, (_, content) => `\${${content}}`)}\`;`,
		`}`,
	].join(EOL);
}

export function createExportFromLanguageAndNamespace(language: string, namespace: string) {
	return `export * as ${namespace} from "${prefixes[0]}${language}/${namespace}";`;
}

export function createRegistry(defaultLanguage: string, languages: string[], namespaces: string[]) {
	return [
		`export const defaultLanguage = ${JSON.stringify(defaultLanguage)} as const;`,
		`export const languages = ${JSON.stringify(languages)} as const;`,
		`export const namespaces = ${JSON.stringify(namespaces)} as const;`,
		"",
		`const modules = {`,
		languages.flatMap(language => [
			`\t${language}: {`,
			namespaces
				.map(namespace => `\t\t\t${namespace}: async () => import("${prefixes[1]}${language}/${namespace}"),`)
				.join(EOL),
			`\t},`,
		]).join(EOL),
		`}`,
		"",
		`export async function importTranslationModule(language: string, namespace: string) {`,
		`\treturn modules[language][namespace]();`,
		`}`,
	].join(EOL);
}
