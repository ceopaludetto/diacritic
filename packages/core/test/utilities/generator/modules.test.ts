/* eslint-disable no-template-curly-in-string */
import type { Entry } from "~/utilities/types";

import { describe, expect, it } from "vitest";

import { createExportFromLanguageAndNamespace, createFunctionFromEntry, createRegistry } from "~/utilities/generator/modules";

describe("createFunctionFromEntry", () => {
	it("should create a function from an entry correctly", () => {
		const entry: Entry = {
			name: "hello",
			path: "hello",
			args: [{ name: "name", type: "string" }],
			return: "Hello, ${name}!",
		};

		expect(createFunctionFromEntry(entry)).toMatchSnapshot();
	});
});

describe("createExportFromLanguageAndNamespace", () => {
	it("should create a export from language and namespace correctly", () => {
		expect(createExportFromLanguageAndNamespace("en", "common")).toMatchSnapshot();
	});
});

describe("createRegistry", () => {
	it("should create a registry correctly", () => {
		expect(createRegistry("en", ["en", "pt"], ["common", "zod"])).toMatchSnapshot();
	});

	it("should create a registry for react-native correctly", () => {
		expect(createRegistry("en", ["en", "pt"], ["common", "zod"], true)).toMatchSnapshot();
	});
});
