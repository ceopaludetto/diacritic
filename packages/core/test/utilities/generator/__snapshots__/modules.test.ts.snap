// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`createFunctionFromEntry should create a function from an entry correctly 1`] = `
"export function hello(name: string) {
	return \`Hello, { name : string }!\`;
}"
`;

exports[`createExportFromLanguageAndNamespace should create a export from language and namespace correctly 1`] = `"export * as common from "/~translations/en/common";"`;

exports[`createRegistry should create a registry correctly 1`] = `
"export const defaultLanguage = "en" as const;
export const languages = ["en","pt"] as const;
export const namespaces = ["common","zod"] as const;

const modules = {
	en: {
			common: async () => import("~translations/en/common"),
			zod: async () => import("~translations/en/zod"),
	},
	pt: {
			common: async () => import("~translations/pt/common"),
			zod: async () => import("~translations/pt/zod"),
	},
}

export async function importTranslationModule(language: string, namespace: string) {
	return modules[language][namespace]();
}"
`;
